File: binarytreepreordertraversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        if not root:
            return res
        stack = []
        while root or stack:
            if root:
                res.append(root.val)
                stack.append(root)
                root = root.left
            else:
                t = stack.pop()
                root = t.right
        return res

File: cpp/binarytreepreordertraversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        if (!root) return vector<int> {};
        stack<TreeNode*> s;
        vector<int> res;
        s.push(root);
        while (!s.empty()) {
            auto target = s.top();
            s.pop();
            res.push_back(target->val);
            if (target->right) s.push(target->right);
            if (target->left) s.push(target->left);
        }
        return res;
    }
};
